### DISCUSSION TITLE
Create a tone, play it back and record it simultaneously

### DISCUSSION THREAD

**Discussion by Adazus** [WebAudio/web-audio-api#2619] [Q&A] [answered]
Hello community,

I have been looking for a solution to the following problem for some time now:
I am not able to:
- generate a sine tone
- output it via the speakers 
and **at the same time** **(this is the issue)**
- start a recording via microphone and record the (emitted) sine tone of the oscillator.

In other words:
All ambient sounds are recorded by the microphone except the frequency generated by the oscillator reproduced by the speaker.

General question:
Is it with web-audio-api basically possible to generate a frequency, output it via the loudspeaker and record it by microphone at the same time?

On the demo from “borismus” the issue is visible:
https://borismus.github.io/spectrogram/
The tone generated by clicking on the screen (red line) is not displayed in the spectrogram.

Many thanks in advance.
Best regards
Adazus

**Comment by Adazus** [comment, marked as answer]
Hello together, 
in the meantime, I have received help at slack.com. 
The issue is that when using:
`getUserMedia({audio: true})`
audio-pre-processing is standardly activated.
To switch off pre-processing you can use: 
`getUserMedia({audio: {autoGainControl: false, echoCancellation: false, noiseSuppression: false} })`
BR Adazus

**Comment by mjwilson-google** [comment]
These constraints are part of the Media Capture and Streams spec:
https://www.w3.org/TR/mediacapture-streams/#media-track-supported-constraints

The [MediaStreamAudioDestinationNode](https://webaudio.github.io/web-audio-api/#MediaStreamAudioDestinationNode) definition already links to Media Capture and Streams for the definition of `getUserMedia()`, so I think there's not any spec work necessary regarding this.

I'm glad that you got help on Slack, and thanks for posting the solution.  I will move this issue to the discussion board, so that it can potentially help others with the same problem.
